src/Core/Grid/Presenter/Column/ColumnDataPresenterChain.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Presenter\Column;
+
+use PrestaShop\PrestaShop\Core\Grid\Column\ColumnInterface;
+use RuntimeException;
+
+final class ColumnDataPresenterChain implements ColumnDataPresenterChainInterface
+{
+    /**
+     * @var ColumnDataPresenterInterface[]
+     */
+    private $columnDataPresenters = [];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function present(ColumnInterface $column, array $record, $gridId)
+    {
+        foreach ($this->columnDataPresenters as $columnDataPresenter) {
+            if ($columnDataPresenter->supports($column)) {
+                return $columnDataPresenter->present($column, $record, $gridId);
+            }
+        }
+
+        throw new RuntimeException(sprintf('Data presenter for column "%s" does not exist.', get_class($column)));
```

Comment:
@matks i'm not sure if we have can have default presenter or if it even makes sense to have it?

Reply:
Even if it returns an empty array, I want to make sure that grid works out-of-the-box if a developer adds a new column that does not need any behavior modification for twig template and presenter.

Reply:
Agree, but it shouldn't be an empty array, it should be the 1.7.5 behavior: open by default and if you need to control the workflow: hooks for them, presenter (with an hook) for us.

Reply:
> I want to make sure that grid works out-of-the-box if a developer adds a new column

This is how I understand it:

**Before:** Column definition + Row data + template with presenter (yes, in 1.7.5 template had 2 responsibilities: 1. figure out how & what data should be displayed by combining Column definition, it's options and row data 2. display data). _This way we are passing Column definition object & whole row data to template._

**After this PR:** Column data presenter (accepts Column definition & row data and figures out how to present it) + template. This way, we pass only final data (no objects) to template.

@matks combining these 2 approaches into single one, would make maintenance hard. In addition, we will never know what data should be pass to template without presenter. :/

> it shouldn't be an empty array, it should be the 1.7.5 behavior

I think it's a bad idea. We want to make an improvement and we have chance to do it clean. Why not making this improvement separately and deprecate existing behavior? It seems that we want to follow old development path from 1.6, 1.5 & etc versions - build new features on top of existing one's, make it more complex and harder to maintain. Later, if another improvement is needed, we will put it on top again, thinking it's easier for developers, but in fact, it's not.

Let me know what you think.

Reply:
> I think it's a bad idea. We want to make an improvement and we have chance to do it clean. Why not making this improvement separately and deprecate existing behavior? It seems that we want to follow old development path from 1.6, 1.5 & etc versions - build new features on top of existing one's, make it more complex and harder to maintain. Later, if another improvement is needed, we will put it on top again, thinking it's easier for developers, but in fact, it's not.
> 
> Let me know what you think.

Yes it will be harder to maintain, that is for sure. But this is the price we pay to preserve backward compatibility. It's the same price Wordpress pays when it fixes bugs in Wordpress v4 although the latest is v5, it's the same price Symfony paid for last v2 versions while preparing the work for v3. You must maintain the deprecated code and the new code until next major version. It is paid in developer time (which is loooot of money), code quantity (lines of code + number of files), complexity, and even response time. But this promise we make to prestashop developers is what keeps them using the software.

Also this new Presenter is probably a prolific concept that could generate multiple features and benefits for the next prestashop versions: faster rendering because of optimization, high-level customization capabilities ... and I think we will not be able to address it all in 1.7.6 . So the plan is to introduce it in 1.7.6, flag it as `@internal` in order for people to be warned that this is experimental stuff and it will likely evolve in 1.7.7, and refine and polish it in 1.7.7 to address all our needs. So it's quite early to deprecate the previous system.

Reply:
> You must maintain the deprecated code and the new code until next major version.

What I meant is no to remove any code, but have 2 ways of rendering grid: 1. Current way (deprecate it and keep untill next major) 2. New way (with column presenters)

That ofcourse would require to have two different templates: 1. For rendering grid old way (with objects in templates & logic inside them). 2. For rendering grid without logic in templates because presenters takes care of that. 

However, from what I understand, you suggest to have 2 Grid presenters **but single template for both**, right? 

> So the plan is to introduce it in 1.7.6, flag it as `@internal` in order for people to be warned that this is experimental stuff

I totally agree with that, it's nearly impossible to make stable software in single iteration. What I wanted to do is mark whole Grid as `@experimental` for 1.7.5 and keep it for a few minor releases, so we could polish it. I remeber having discussion about it with Mickael or Pablo, but the decision was not to do so, can't really remember why. :/

Reply:
> However, from what I understand, you suggest to have 2 Grid presenters **but single template for both**, right?

Yes indeed. Having 2 sets of templates looks like a big drawback to this idea 😞